### PR TITLE
4.x

### PR DESCRIPTION
**Pull request by vladimirpotekhin**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

update to 15 angular

# **WARNING** cut 3.x before merge

### CODE REVIEW

File: projects/demo/src/utils/load-assets.ts [hunk 1]
```diff
@@ -1,6 +1,3 @@
 export function assets(path: TemplateStringsArray): string {
-    const dirname = new URL(`.`, import.meta.url);
-    const relativePath = `../assets${path.join(``)}`;
-
-    return new URL(relativePath, dirname).toString();
+    return `./assets/${path.join(``)}`;
```
Comment (waterplea) [inline projects/demo/src/utils/load-assets.ts:2]: What's up with this change?

File: tsconfig.json [hunk 2]
```diff
@@ -5,6 +5,9 @@
     "compilerOptions": {
         "baseUrl": "./",
         "outDir": "./dist",
+        "useDefineForClassFields": false,
+        "target": "ES2022",
```
Comment (waterplea) [inline tsconfig.json:9]: Is that not too modern?

**Review by waterplea — APPROVED** [review]

File: projects/demo/src/utils/load-assets.ts (hunk 1 above)
Comment (vladimirpotekhin) [reply projects/demo/src/utils/load-assets.ts:2]: for some reason the previous version returned paths like `file:///Users/.../taiga-ui/projects/demo/src/assets/images/...` which caused the error `Not allowed to upload local resource`

File: package.json [hunk 3]
```diff
@@ -99,33 +99,36 @@
...
+        "@angular/cli": "15.2.10",
+        "@angular/compiler-cli": "15.2.10",
+        "@angular/language-service": "15.2.10",
         "@commitlint/cli": "18.4.3",
-        "@nx/jest": "17.2.1",
-        "@nx/node": "17.2.1",
-        "@nx/workspace": "17.2.1",
+        "@cypress/webpack-batteries-included-preprocessor": "^3.0.2",
+        "@cypress/webpack-preprocessor": "^6.0.0",
+        "@nx/angular": "^17.2.0",
```
Comment (nsbarsukov) [inline package.json:128]: Use fixed versions

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@taiga-ui/components",
-    "version": "3.58.0",
+    "version": "4.0.0",
```
Comment (nsbarsukov) [inline package.json:3]:
```
npm run release:major
```

will do the following:
Screenshot 2023-12-12 at 18 09 50

File: package-lock.json
[the remark this answers is not in the document]
Comment (nsbarsukov) [reply package-lock.json:1]: 1. Try to clean all `node_modules` (root and all nested ones)
2. Delete `package-lock.json`
3. Run `npm i --workspaces --include-workspace-root`

File: projects/demo/package.json [hunk 5]
```diff
@@ -9,10 +9,11 @@
         "@tinkoff/ng-dompurify": "4.0.0",
         "date-fns": "2.30.0",
         "dompurify": "3.0.6",
-        "ngx-markdown": "10.1.1"
+        "marked": "^4.0.17",
+        "ngx-markdown": "15.1.2"
     },
     "devDependencies": {
-        "@nguniversal/builders": "12.1.3",
+        "@nguniversal/builders": "15.2.1",
```
Comment (nsbarsukov) [inline projects/demo/package.json:16]: Should we also bump `@nguniversal/express-engine` ?

File: projects/demo/package.json [hunk 6]
```diff
@@ -9,10 +9,11 @@
         "@tinkoff/ng-dompurify": "4.0.0",
         "date-fns": "2.30.0",
         "dompurify": "3.0.6",
-        "ngx-markdown": "10.1.1"
+        "marked": "^4.0.17",
```
Comment (nsbarsukov) [inline projects/demo/package.json:12]: fixate it

File: projects/demo/project.json [hunk 7]
```diff
@@ -14,7 +14,7 @@
             }
         },
         "build": {
-            "executor": "@angular-builders/custom-webpack:browser",
+            "executor": "@nx/angular:webpack-browser",
```
Comment (nsbarsukov) [inline projects/demo/project.json:17]: Should we also change executor for `server`-target ?

https://nx.dev/nx-api/angular/executors/webpack-server

File: projects/testing/utils/input/cleaner.unit-common.ts [hunk 8]
```diff
@@ -17,7 +17,7 @@ interface TestParams {
 
 interface TestComponent {
     cleaner: boolean;
-    control?: FormControl;
+    control?: UntypedFormControl;
```
Comment (nsbarsukov) [inline projects/testing/utils/input/cleaner.unit-common.ts:20]: Should we revert this line ?

File: package.json (hunk 4 above)
Comment (vladimirpotekhin) [reply package.json:3]: for now this is necessary to create snapshots for e2e

Comment (vladimirpotekhin) [reply package.json]: although no other PRs will be sent to this PR, reverted to 3.58

File: projects/demo/project.json (hunk 7 above)
Comment (vladimirpotekhin) [reply projects/demo/project.json:17]: there are some problems with nx executor, let's do it in a separate pr

**Review by nsbarsukov — APPROVED** [review]

File: projects/core/internal/primitive-year-month-pagination/test/primitive-year-month-pagination.component.spec.ts [hunk 9]
```diff
@@ -59,10 +59,6 @@ describe(`PrimitiveYearMonthPaginationComponent`, () => {
             expect(component.value).toBe(testComponent.value);
         });
 
-        it(`min`, () => {
```
Comment (splincode) [inline projects/core/internal/primitive-year-month-pagination/test/primitive-year-month-pagination.component.spec.ts:62]: why?

File: projects/demo/src/modules/components/select/examples/11/index.ts [hunk 10]
```diff
@@ -1,5 +1,5 @@
 import {Component} from '@angular/core';
-import {FormControl} from '@angular/forms';
+import {UntypedFormControl} from '@angular/forms';
```
Comment (splincode) [inline projects/demo/src/modules/components/select/examples/11/index.ts:2]: why not just

```ts
import {UntypedFormControl as FormControl} from '@angular/forms';
```

everywhere

File: projects/experimental/components/toggle/toggle.component.ts [hunk 11]
```diff
@@ -28,7 +29,7 @@ import {TUI_TOGGLE_OPTIONS, TuiToggleOptions} from './toggle.options';
 })
 export class TuiToggleComponent {
     @Input()
-    size = this.options.size;
+    size: TuiSizeS = this.options.size;
```
Comment (splincode) [inline projects/experimental/components/toggle/toggle.component.ts:32]: why?

**Review by splincode — APPROVED** [review]

**Pull request merged**